Configure DNS using terraform

tlwr opened the pull request:
~~Kops has a domain name of which it is aware. Now that we have a domain
that we can use we can use this. This should simplify our service
configuration when using ingress, etc.~~

~~Additionally we were managing the state of the DNS records using
scripting, when we can use terraform instead.~~



    Instead of doing route53 stuff using a script, we can do it using
    terraform when we create the cluster, we can then use terraform to pull
    out the IDs

    The side effect of this is we are reducing the number of IAM permissions
    required to converge a cluster. This means that we can have separate
    access levels for creating and converging a cluster, as the cluster
    manages its own IAM.

LeePorte commented:
Upon testing I am seeing the following

```
Apply complete! Resources: 10 added, 0 changed, 0 destroyed.

Outputs:

azs = eu-west-2a,eu-west-2b,eu-west-2c
domain_name = longboy.govsvc.uk
domain_zone_id = Z2QL2CH69SKVAI
subnet_ids = subnet-0d1a082a78eea3d1f,subnet-0c2c121995f50e437,subnet-050484311acfff9cb
vpc_id = vpc-00b7723fdbde3bca2
/Users/leeporte/go-code/src/github.com/alphagov/re-boot/kops
vpc-00b7723fdbde3bca2
subnet-0d1a082a78eea3d1f,subnet-0c2c121995f50e437,subnet-050484311acfff9cb
eu-west-2a,eu-west-2b,eu-west-2c
longboy.govsvc.uk
I0817 13:28:26.521125   14774 create_cluster.go:1318] Using SSH public key: /Users/leeporte/.ssh/id_rsa.pub

error doing DNS lookup for NS records for "longboy.govsvc.uk": lookup longboy.govsvc.uk on [2a02:8010:62bb::227:eff:fe16:bd48]:53: no such host
```
It looks like you'll need to grab the name servers from the newly created hosted zone and populate an appropriate NS record in the parent hosted zone.

LeePorte reviewed (changes requested):
Doesn't apply at present

tlwr commented:
When I run
```
./create-cluster longboy
```

I get
```
Terraform will perform the following actions:

  ~ aws_internet_gateway.igw
      tags.Name:                                    "longboy.k8s.local" => "longboy.govsvc.uk"

  + aws_route53_zone.domain
      id:                                           <computed>
      comment:                                      "Managed by Terraform"
      force_destroy:                                "false"
      name:                                         "longboy.govsvc.uk"
      name_servers.#:                               <computed>
      vpc_region:                                   <computed>
      zone_id:                                      <computed>

  ~ aws_route_table.rt
      tags.Name:                                    "longboy.k8s.local" => "longboy.govsvc.uk"

  ~ aws_subnet.subnets[0]
      tags.Name:                                    "longboy.k8s.local-0" => "longboy.govsvc.uk-0"
      tags.kubernetes.io/cluster/longboy.govsvc.uk: "" => "shared"
      tags.kubernetes.io/cluster/longboy.k8s.local: "shared" => ""

  ~ aws_subnet.subnets[1]
      tags.Name:                                    "longboy.k8s.local-1" => "longboy.govsvc.uk-1"
      tags.kubernetes.io/cluster/longboy.govsvc.uk: "" => "shared"
      tags.kubernetes.io/cluster/longboy.k8s.local: "shared" => ""

  ~ aws_subnet.subnets[2]
      tags.Name:                                    "longboy.k8s.local-2" => "longboy.govsvc.uk-2"
      tags.kubernetes.io/cluster/longboy.govsvc.uk: "" => "shared"
      tags.kubernetes.io/cluster/longboy.k8s.local: "shared" => ""

  ~ aws_vpc.vpc
      tags.Name:                                    "longboy.k8s.local" => "longboy.govsvc.uk"
```

How did you run the apply?

LeePorte commented:
I used the same method, it fails after answering yes. Give it a try with shortboy.

tlwr commented:
Works on my machine (tm)

Commit:
{short,medium,farm}boy parity /w longboy

Commit:
Configure DNS using terraform

Instead of doing route53 stuff using a script, we can do it using
terraform when we create the cluster, we can then use terraform to pull
out the IDs

The side effect of this is we are reducing the number of IAM permissions
required to converge a cluster. This means that we can have separate
access levels for creating and converging a cluster, as the cluster
manages its own IAM.
